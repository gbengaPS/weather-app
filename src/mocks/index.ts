export const citiesMock = {
  base: "stations",
  clouds: {all: 0},
  cod: 200,
  coord: {lon: 116.4, lat: 39.91},
  dt: 1602226045,
  id: 1816670,
  main: {temp: 23.52, feels_like: 22.22, temp_min: 23.33, temp_max: 23.89, pressure: 1014, humidity: 47},
  name: "Beijing",
  sys: {type: 3, id: 2000022, country: 'CN', sunrise: 1602195502, sunset: 1602236679},
  timezone: 28800,
  visibility: 10000,
  weather: [{id: 800, main: "Clear", description: "clear sky", icon: "01d"}],
  wind: {speed: 2.55, deg: 209},
};

export const initialStateMock =  {
  favorites: { 'Beijing': { data: citiesMock } },
  weatherInfo: [],
  notes: { 'Beijing': 'hello world' },
  isLoading: false
}