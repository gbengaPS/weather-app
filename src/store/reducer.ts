const reducer = (state: any, action: any) => {
  return {}
}

export default reducer;