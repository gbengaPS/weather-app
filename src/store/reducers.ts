import { REMOVE_WEATHER_INFO, ADD_WEATHER_INFO, ADD_FAVORITE, REMOVE_FAVORITE } from '../constants';
import { FavoritesMap, WeatherInfo } from '../types';
import { saveFavorites } from '../helpers';

export const weatherInfoReducer = (state: WeatherInfo[], action: any) => {
  switch(action.type) {
    case REMOVE_WEATHER_INFO:
      const filteredInfo = state.filter((info: WeatherInfo) => {
        return info.data.name !== action.cityName;
      });
      localStorage.setItem('largestCitiesInfo', JSON.stringify(filteredInfo))
      return filteredInfo;

    case ADD_WEATHER_INFO:
      return [ ...state, ...action.weatherInfo ];
    default:
      return state;
  }
}

export const favoritesReducer = (state: FavoritesMap, action: any) => {
  switch(action.type) {
    case ADD_FAVORITE:
      const newState = { ...state, ...action.favoriteItem };
      saveFavorites(newState)
      return newState;

    case REMOVE_FAVORITE:
      const stateCopy = {...state};
      delete stateCopy[action.cityName];
      saveFavorites(stateCopy);
      return stateCopy;
    default:
      return state
  }
}