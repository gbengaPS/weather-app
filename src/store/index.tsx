import React, { createContext, useReducer } from 'react';
import { FavoritesMap, WeatherInfo } from '../types';
import { weatherInfoReducer, favoritesReducer } from './reducers';

interface InitialStateType {
	favorites: FavoritesMap
	weatherInfo: WeatherInfo[]
}

const initialState  = {
	favorites: {},
	weatherInfo: []
}

const AppContext = createContext<{
    state: InitialStateType;
    dispatch: React.Dispatch<any>;
  }>({
    state: initialState,
    dispatch: () => null
  });

const mainReducer = (state: InitialStateType, action: any) => ({
  weatherInfo: weatherInfoReducer(state.weatherInfo, action),
  favorites: favoritesReducer(state.favorites, action)
});

const AppProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(mainReducer, initialState);

  return (
      <AppContext.Provider value={{state, dispatch}}>
      {children}
      </AppContext.Provider>
  )
}
  
  export { AppContext, AppProvider };