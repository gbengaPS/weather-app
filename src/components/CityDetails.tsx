import React, { useContext } from 'react';
import { useParams, Link } from 'react-router-dom';
import {
  getDateFromTime,
  getIconURL,
  getLocalTime,
  usePopulateStore
} from '../helpers';
import { AppContext } from '../store';
import '../styles/cityDetails.scss';
import { WeatherInfo } from '../types';
import Notes from './Notes';

const CityDetails = () => {
  const { state } = useContext(AppContext);
  usePopulateStore();

  const { cityName } = useParams<{ cityName:string }>();
  const userCurrentCity = JSON.parse(localStorage.getItem('userCurrentCity') || '{}');
  
  let cityDetails : WeatherInfo | undefined = state.favorites[cityName];


  if (state.isLoading) {
    return <img src="/loader.gif"  alt="loading indicator"/>
  }

  if (userCurrentCity?.data.name === cityName) {
    cityDetails = userCurrentCity;
  }

  if(!cityDetails) {
    cityDetails = state.weatherInfo.find((info) => {
      return info.data.name === cityName;
    })
  }

  const getCityDetails = () => {
    if (!cityDetails) {
      return <p>City not found</p>;
    }

    const { data } = cityDetails;
    const weatherDescription = data.weather[0].description;
    return (
      <div>
        <div className="details-header">
          <p>
            {cityDetails.data.name}
          </p>
          <p>{getDateFromTime(data.sys.sunrise)}</p>
        </div>
        <div className="city-details-card">
          <div className="space-between">
            <div className="vertical-align">
              <p className="details-temp">{Math.floor(data.main.temp)}&#730;</p>
              <p>Feels like: {Math.floor(data.main.feels_like)}&#730;</p>
            </div>

            <div className="vertical-align">
              <img src={getIconURL(data.weather[0].icon)} alt={weatherDescription} />
              <p>{weatherDescription}</p>
            </div>
          </div>
          <div className="space-between">
            <p>Sunrise: {getLocalTime(data.sys.sunrise, data.timezone)} AM</p>
            <p>Sunset: {getLocalTime(data.sys.sunset, data.timezone)} PM</p>
          </div>
        </div>

        <ul className="details-list">
          <li><span>Max Temp:</span> <span>{data.main.temp_max}&#730;</span></li>
          <li><span>Min Temp:</span> <span>{data.main.temp_min}&#730;</span></li>
          <li><span>Humidity:</span> <span>{data.main.humidity}%</span></li>
          <li><span>Pressure:</span> <span>{data.main.pressure} hPa</span></li>
          <li><span>Wind Speed: </span> <span>{data.wind.speed} m/s</span></li>
          <li><span>Wind Deg:</span> <span>{data.wind.deg}&#730;</span></li>
        </ul>
      </div>
    )
  }

  return (
    <div className="city-details-wrapper app">
      <Link to="/"> Go Back</Link>
      {getCityDetails()}
      <Notes cityName={cityName} />
    </div>
  )
};


export default CityDetails;
