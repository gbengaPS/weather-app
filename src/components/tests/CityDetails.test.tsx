import React from 'react';
import { render } from '@testing-library/react';
import CityDetails from '../CityDetails';
import { initialStateMock } from '../../mocks';
import { AppContext} from '../../store';

jest.mock('../Notes', () => jest.fn()
.mockImplementation(() => <>**Notes**</>))

jest.mock('react-router-dom', () => ({
  useParams: () => { return { cityName: 'Beijing' } },
  Link: () => <>**Link**</>
}))

describe('CityDetails', () => {

  const dispatch = jest.fn();

  it('should match snapshot', () => {
    const { container } = render(
      <AppContext.Provider value={{ state: initialStateMock, dispatch}}>
        <CityDetails />
      </AppContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when city cannot be found', () => {
    initialStateMock.favorites = {}
    const { container } = render(
      <AppContext.Provider value={{ state: initialStateMock, dispatch}}>
        <CityDetails />
      </AppContext.Provider>
    );
    expect(container).toMatchSnapshot();
  })
});