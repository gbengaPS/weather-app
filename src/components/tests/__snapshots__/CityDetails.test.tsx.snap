// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityDetails should match snapshot when city cannot be found 1`] = `
<div>
  <div
    class="city-details-wrapper app"
  >
    **Link**
    <p>
      City not found
    </p>
  </div>
</div>
`;
